core/src/tile/tileManager.cpp

```diff
@@ -3,237 +3,350 @@
+
+        for (auto& worker : m_workers) {
+            if (worker->isRunning()) { continue; }
+
+            worker->process(m_scene->getStyles());
+        }
     }
-    
-    const std::set<TileID>& visibleTiles = m_view->getVisibleTiles();
-    
-    // Loop over visibleTiles and add any needed tiles to tileSet
-    {
-        auto setTilesIter = m_tileSet.begin();
-        auto visTilesIter = visibleTiles.begin();
-        
-        while (visTilesIter != visibleTiles.end()) {
-            
-            if (setTilesIter == m_tileSet.end() || *visTilesIter < setTilesIter->first) {
-                // tileSet is missing an element present in visibleTiles
-                addTile(*visTilesIter);
+
+    if (!m_readyTiles.empty()) {
+        std::lock_guard<std::mutex> lock(m_readyTileMutex);
+        auto it = m_readyTiles.begin();
+
+        while (it != m_readyTiles.end()) {
+            auto& task = *it;
+            auto& tile = *(task->tile);
+
+            if (setTileState(tile, MapTile::Ready)) {
+                cleanProxyTiles(tile);
                 m_tileSetChanged = true;
-                ++visTilesIter;
-            } else if (setTilesIter->first < *visTilesIter) {
-                // visibleTiles is missing an element present in tileSet (handled below)
-                ++setTilesIter;
-            } else {
-                // tiles in both sets match, move on
-                ++setTilesIter;
-                ++visTilesIter;
             }
+
+            it = m_readyTiles.erase(it);
         }
     }
-    
-    // Loop over tileSet and remove any tiles that are neither visible nor proxies
-    {
-        auto setTilesIter = m_tileSet.begin();
-        auto visTilesIter = visibleTiles.begin();
-        
-        while (setTilesIter != m_tileSet.end()) {
-            
-            if (visTilesIter == visibleTiles.end() || setTilesIter->first < *visTilesIter) {
-                // visibleTiles is missing an element present in tileSet
-                if (setTilesIter->second->getProxyCounter() <= 0) {
-                    removeTile(setTilesIter);
-                    m_tileSetChanged = true;
+
+    if (m_view->changedOnLastUpdate() || m_tileSetChanged) {
+
+        const std::set<TileID>& visibleTiles = m_view->getVisibleTiles();
+
+        // Loop over visibleTiles and add any needed tiles to tileSet
+        {
+            auto setTilesIter = m_tileSet.begin();
+            auto visTilesIter = visibleTiles.begin();
+
+            while (visTilesIter != visibleTiles.end()) {
+
+                auto& visTileId = *visTilesIter;
+                auto& curTileId = setTilesIter == m_tileSet.end() ? NOT_A_TILE : setTilesIter->first;
+
+                if (visTileId == curTileId) {
+                    if (setTilesIter->second->state() == MapTile::None) {
+                        logMsg("[%d, %d, %d] - Enqueue\n", curTileId.z, curTileId.x, curTileId.y);
+                        m_loadQueue.push_back(curTileId);
```

Comment:
That's because we want tileid to be immutable. It's used as a key in
std::map.
On Jun 29, 2015 5:58 PM, "Hannes Janetzek" <notifications@github.com> wrote:

> In core/src/tile/tileManager.cpp
> <https://github.com/tangrams/tangram-es/pull/161#discussion_r33518595>:
>
> > +        const std::set<TileID>& visibleTiles = m_view->getVisibleTiles();
> > +
> > +        // Loop over visibleTiles and add any needed tiles to tileSet
> > +        {
> > +            auto setTilesIter = m_tileSet.begin();
> > +            auto visTilesIter = visibleTiles.begin();
> > +
> > +            while (visTilesIter != visibleTiles.end()) {
> > +
> > +                auto& visTileId = *visTilesIter;
> > +                auto& curTileId = setTilesIter == m_tileSet.end() ? NOT_A_TILE : setTilesIter->first;
> > +
> > +                if (visTileId == curTileId) {
> > +                    if (setTilesIter->second->state() == MapTile::None) {
> > +                        logMsg("[%d, %d, %d] - Enqueue\n", curTileId.z, curTileId.x, curTileId.y);
> > +                        m_loadQueue.push_back(curTileId);
>
> Actually loadQueue using a std::list since vector (or deque) would not
> allow to sort TileIDs. Which reminds me to ask why TileID needs to have
> const members?
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/tangrams/tangram-es/pull/161/files#r33518595>.
>